### PR TITLE
Workaround container networking limitations

### LINKED ISSUE #1197: AddReference between containers
Hey

I'm trying to connect to container together with `WithReference`. But I'm not able to get the AllocatedEndpointAnnotation. 

Here is a very simple example. (I'm aware that this would not make a lot of sense ;) )

```
var redis = builder.AddRedisContainer("redis");
var rabbitMq = builder.AddRabbitMQContainer("rabbitmq");

rabbitMq.WithReference(redis);
```

When I start the application locally on my mac. I get the error:  `Aspire.Hosting.DistributedApplicationException: Redis resource does not have endpoint annotation.`

This is the same behavior when I try to reference my custom containers.

### PR DESCRIPTION
**Pull request by davidfowl**
- Until we setup container networking, make sure containers can talk to other containers when using WithReference. This won't work when using raw strings/urls.

Preview2 mitigation for https://github.com/dotnet/aspire/issues/128

Fixes https://github.com/dotnet/aspire/issues/1197

### CODE REVIEW

**Comment by mitchdenny** [comment]
Taking bets on how long this work around will hang around for?

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs [hunk 1]
```diff
@@ -69,9 +69,9 @@ public async Task RunApplicationAsync(CancellationToken cancellationToken = defa
             PrepareServices();
             PrepareContainers();
             PrepareExecutables();
-            PrepareProxylessServices();
+            // PrepareProxylessServices();
 
-            await CreateContainerSingletonsAsync(cancellationToken).ConfigureAwait(false);
+            // await CreateContainerSingletonsAsync(cancellationToken).ConfigureAwait(false);
```
Comment (davidfowl) [inline src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:75]: @bwateratmsft @karolz-ms I ended up disabling this to make the PR work. Any guidance here on how this is supposed to work would be appreciated.

File: src/Aspire.Hosting/Utils/HostNameResolver.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    internal static string ReplaceLocalhostWithContainerHost(string value, IConfiguration configuration)
+    {
+        // https://stackoverflow.com/a/43541732/45091
+
+        // This configuration value is a workaround for the fact that host.docker.internal is not available on Linux by default.
+        var hostName = configuration["AppHost:ContainerHostname"] ?? "host.docker.internal";
+
+        // REVIEW: Handle IPV6?
+        return value.Replace("localhost", hostName, StringComparison.OrdinalIgnoreCase)
+                    .Replace("127.0.0.1", hostName);
```
Comment (karolz-ms) [inline src/Aspire.Hosting/Utils/HostNameResolver.cs:21]: Aspire uses `localhost` by default, but the AppHost code is free to change that. I would add the IPV6 `[::1]` to the replacement list. 

Other IPv4 loopback addresses (not `127.0.0.1`) will not work, but they are probably very unlikely to be used, and we are going to rework this in next release.

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs (hunk 1 above)
Comment (karolz-ms) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:75]: We had a conversation about this in https://github.com/dotnet/aspire/pull/886. Long story short, we need more time and design to enable this.

**Review by karolz-ms — APPROVED** [review]

**Comment by davidfowl** [comment]
cc @eerhardt This change disables proxyless containers for preview2 until we have more runway to fix the other issues

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs (hunk 1 above)
Comment (JamesNK) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:80]: FYI unused code was causing code analysis warnings. I've removed it here - https://github.com/dotnet/aspire/pull/1286. Source is still available in the file history 😄

Comment (karolz-ms) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:80]: Thank you!

**Pull request merged**